File: typespec/hub/auth/password.ts
import type { NewPassword } from "../../common/authentication.ts";
import { isNewPassword, isOpaqueToken } from "../../common/authentication.ts";
import type { EmailAddress } from "../../common/common.ts";
import { isEmailAddress, normalizeEmailAddress } from "../../common/common.ts";

export type HubPasswordResetToken = string;

export interface RequestPasswordResetRequest {
  email_address: EmailAddress;
}

export function normalizeRequestPasswordResetRequest(
  request: RequestPasswordResetRequest,
): RequestPasswordResetRequest {
  return {
    ...request,
    email_address: normalizeEmailAddress(request.email_address),
  };
}

export function validateRequestPasswordResetRequest(
  request: RequestPasswordResetRequest,
): string[] {
  return isEmailAddress(normalizeEmailAddress(request.email_address))
    ? []
    : ["email_address"];
}

export interface CompletePasswordResetRequest {
  reset_token: HubPasswordResetToken;
  new_password: NewPassword;
}

export function validateCompletePasswordResetRequest(
  request: CompletePasswordResetRequest,
): string[] {
  const fields: string[] = [];
  if (!isOpaqueToken(request.reset_token)) fields.push("reset_token");
  if (!isNewPassword(request.new_password)) fields.push("new_password");
  return fields;
}

export interface ChangePasswordRequest {
  new_password: NewPassword;
}

export function validateChangePasswordRequest(
  request: ChangePasswordRequest,
): string[] {
  return isNewPassword(request.new_password) ? [] : ["new_password"];
}

File: typespec/hub/auth/password.go
package auth

import "github.com/vetchium/src/typespec/common"

type HubPasswordResetToken common.OpaqueToken

type RequestPasswordResetRequest struct {
	EmailAddress common.EmailAddress `json:"email_address"`
}

func (r *RequestPasswordResetRequest) Normalize() {
	r.EmailAddress = common.NormalizeEmailAddress(r.EmailAddress)
}

func (r RequestPasswordResetRequest) Validate() []string {
	if !common.IsEmailAddress(r.EmailAddress) {
		return []string{"email_address"}
	}
	return []string{}
}

type CompletePasswordResetRequest struct {
	ResetToken  HubPasswordResetToken `json:"reset_token"`
	NewPassword common.NewPassword    `json:"new_password"`
}

func (r *CompletePasswordResetRequest) Normalize() {}

func (r CompletePasswordResetRequest) Validate() []string {
	fields := make([]string, 0, 2)
	if !common.IsOpaqueToken(string(r.ResetToken)) {
		fields = append(fields, "reset_token")
	}
	if !common.IsNewPassword(r.NewPassword) {
		fields = append(fields, "new_password")
	}
	return fields
}

type ChangePasswordRequest struct {
	NewPassword common.NewPassword `json:"new_password"`
}

func (r *ChangePasswordRequest) Normalize() {}

func (r ChangePasswordRequest) Validate() []string {
	if !common.IsNewPassword(r.NewPassword) {
		return []string{"new_password"}
	}
	return []string{}
}
